### PR TITLE
MAINT: Avx512 intrinsics implementation for float64 input np.log

### PR DESCRIPTION
**Pull request by xiegengxin**
This patch implements np.log by using avx512 intrinsics  when input is float64. The algorithm references from: [1] Tang, Ping Tak Peter. Table-lookup algorithms for elementary functions and their error analysis. No. CONF-9106103-1. Argonne National Lab., IL (USA), 1991. and [2] Muller, Jean-Michel. "Elementary functions: algorithms and implementation." (2016).

**Performance**
The performance is tested by some cases under https://github.com/numpy/numpy/blob/f3ff68e5fe874f9626fcbe85000549083f4f222a/benchmarks/benchmarks/bench_avx.py#L25  and https://github.com/numpy/numpy/blob/f3ff68e5fe874f9626fcbe85000549083f4f222a/benchmarks/benchmarks/bench_avx.py#L154

Compare with master branch (scalar version by calling glibc's function), the new implementation has about 6x faster. The speed of train LogisticRegression has about 1.2x than before.

| item | master | avx512-log-float64| ratio|
|------| ---------|---------------------|------|
| bench_avx.LogisticRegression.time_train(<class 'numpy.float64'>)|  4.66±0.01s |      3.87±0.01s   |                    0.83  |
| bench_avx.AVX_UFunc_log.time_log(2, 'd')   |    211±1μs    |   36.7±0.5μs  |                      0.17  |
 |bench_avx.AVX_UFunc_log.time_log(4, 'd')|    212±0.9μs   |    36.2±0.3μs    |                  0.17  |
 |bench_avx.AVX_UFunc_log.time_log(1, 'd')|       212±1μs    |    35.1±1μs         |                0.17  |

**Accuracy**
Tang's paper has analysed the error of his algorithm, it is less than 1ulp.

**module size**
_multiarray_umath.so module has grown about 24 kB after adding this PR.

### CODE REVIEW

File: benchmarks/benchmarks/bench_avx.py [hunk 1]
```diff
@@ -39,6 +39,19 @@ def setup(self, ufuncname, stride, dtype):
     def time_ufunc(self, ufuncname, stride, dtype):
         self.f(self.arr[::stride])
 
+class AVX_UFunc_log(Benchmark):
```
Comment (Qiyu8) [inline benchmarks/benchmarks/bench_avx.py:42]: `np.log` already has a benchmark test in `AVX_UFunc`, Is there any special reason to write a new one?

File: numpy/core/src/umath/npy_simd_data.h [hunk 2]
```diff
@@ -134,4 +134,149 @@ static npy_uint64 EXP_Table_tail[32] = {
 #define NPY_COEFF_INVF7_SINEf -0x1.a06bbap-13f
 #define NPY_COEFF_INVF9_SINEf 0x1.7d3bbcp-19f
 
+/*
+ * Lookup table of log(c_k)
+ */
+static npy_uint64 LOG_TABLE_TOP[64] = {
```
Comment (Qiyu8) [inline numpy/core/src/umath/npy_simd_data.h:140]: Where are these data come from? need to add comment here.

File: numpy/core/src/umath/simd.inc.src [hunk 3]
```diff
@@ -3144,6 +3177,207 @@ AVX512F_exp_DOUBLE(npy_double * op,
 #endif
 #endif
 
+/*
+ * Vectorized implementation of log double using AVX512
+ * 1) if x = 0; return -INF
```
Comment (Qiyu8) [inline numpy/core/src/umath/simd.inc.src:3182]: need add algorithm link

File: numpy/core/src/umath/simd.inc.src [hunk 4]
```diff
@@ -3144,6 +3177,207 @@ AVX512F_exp_DOUBLE(npy_double * op,
...
+            __m512i i = _mm512_and_epi64(_mm512_srai_epi64(tmp, 52 - 6),
+                            _mm512_set1_epi64(0x3fULL));
+            __m512i ik = _mm512_srai_epi64(tmp, 52);
+            __m512d z = (__m512d)(_mm512_sub_epi64(ix, _mm512_and_epi64(tmp,
+                            _mm512_set1_epi64(0xfff0000000000000))));
+            /* c = i/64 + 1 */
+            __m512d c = _mm512_fmadd_pd(_mm512_cvtepi64_pd(i), mInv64, ones_d);
+
+            /* u = 2 * (z - c) / (z + c) */
+            __m512d u = _mm512_div_pd(_mm512_sub_pd(z, c), _mm512_add_pd(z,c));
```
Comment (Qiyu8) [inline numpy/core/src/umath/simd.inc.src:3317]: missing space.

**Review by Qiyu8 — CHANGES_REQUESTED** [review]

File: benchmarks/benchmarks/bench_avx.py (hunk 1 above)
Comment (xiegengxin) [reply benchmarks/benchmarks/bench_avx.py:42]: Yes, the old benchmark using a ones-array as input. This code will call glibc's log function when input equals to 1. Old benchmark can not show out the enhancement of performance. And, glibc implement log in a dispatch when input equals to 1, which is faster than usually.

File: numpy/core/src/umath/simd.inc.src (hunk 3 above)
Comment (xiegengxin) [reply numpy/core/src/umath/simd.inc.src:3182]: Ok, I'll add the algorithm link.

**Comment by sturlamolden** [comment]
How about using GNU vector extensions or autovectorization pragmas instead and leave the specifics of the SIMD dialect to the compiler? 

This code is very specific for Avx512 and will not work on CPUs that do not have this, also it will not work with new SIMD and better extensions in the future. The syntax of Avx512 as with the rest of mmintrin.h is prefix. 

GNU vector extensions work everywhere, except with a certain compiler from Redmond. They are not limited to the Avx512 instruction set. They also provide a nice and readable infix syntax with normal C operators, so we can actually understand what the code is doing. That improves on the maintainability manifold.

We have used GNU extensions for vectorization in e.g. scipy.spatial, with the logic that those who prefer to use Microsoft’s compiler will just have to live with the consequences.

**Comment by mattip** [comment]
> How about using GNU vector extensions or autovectorization pragmas instead and leave the specifics of the SIMD dialect to the compiler?

These kinds of PRs go beyond loop-unrolling and autovectorization.  [NEP 38](https://numpy.org/neps/nep-0038-SIMD-optimizations.html) goes into more detail about the approach we will be taking. Universal intrinsics  work across platforms, and the approach has proven itself useful in end-user projects like OpenCV and GNU Radio, as well as many libraries that provide high-level operations but use intrinsics underneath like [sleef](https://sleef.org) (used by pytorch).

**Comment by sturlamolden** [comment]
> > How about using GNU vector extensions or autovectorization pragmas instead and leave the specifics of the SIMD dialect to the compiler?
> 
> These kinds of PRs go beyond loop-unrolling and autovectorization. [NEP 38](https://numpy.org/neps/nep-0038-SIMD-optimizations.html) goes into more detail about the approach we will be taking. Universal intrinsics work across platforms, and the approach has proven itself useful in end-user projects like OpenCV and GNU Radio, as well as many libraries that provide high-level operations but use intrinsics underneath like [sleef](https://sleef.org) (used by pytorch).

That is exactly my point. There is no use of universal intrinsics in this PR.

**Comment by Qiyu8** [comment]
@sturlamolden Although Universal SIMD is about to merge in few weeks, the value needs to be proven by new implement, and you have to considering the migration cost, so currently we choose to have a solid reference implementation in place, just like the other [SIMD PRs](https://github.com/numpy/numpy/pull/15648).

**Comment by xiegengxin** [comment]
@r-devulap Could you review this codes? Pls.

**Review by Qiyu8 — APPROVED** [review]

**Comment by mattip** [comment]
For completeness, could you rerun the benchmarks and also comapre the size of the shared library before and after this PR?

**Comment by xiegengxin** [comment]
> For completeness, could you rerun the benchmarks and also comapre the size of the shared library before and after this PR?

The performance
|item|     master       |  avx512-log-float64 |ratio|
|--|--|--|--|
|bench_avx.LogisticRegression.time_train(<class 'numpy.float64'>)          |4.63±0s       |3.84±0.01s     |0.83  |
|bench_avx.AVX_UFunc_log.time_log(4, 'd') -         |212±1μs       |37.2±0.2μs     |0.18  |
|bench_avx.AVX_UFunc_log.time_log(2, 'd')       |212±0.4μs     |  36.4±0.1μs  |   0.17 | 
|bench_avx.AVX_UFunc_log.time_log(1, 'd') | 212±0.9μs | 36.1±1μs |  0.17 | 

Still about 6x.

The size of shared library has grown about 23kB.

**Review by mattip — APPROVED** [review]

**Comment by mattip** [comment]
Will give a little more time for other reviews. @r-devulap, @seiko2plus ko2plus: any thoughts?

**Comment by sturlamolden** [comment]
How does this compare to just using vector math libraries, e.g. when compiling with MKL, Accelerate, ACML, etc.? It seems it duplicates a tiny portion of what Intel, Apple, AMD engineers have already implemented (not just for AVX). A vector math solution would also work for any trancendental function, not justb the logarithm. IIRC, Numexpr already does this with MKL.

**Comment by mattip** [comment]
For good and for bad, we chose to implement the ufunc SIMD routines ourselves rather than to re-use libraries listed in [NEP 38](https://numpy.org/neps/nep-0038-SIMD-optimizations.html#related-work). A future NEP may propose to rethink this decision, but would need to take into account the functionality needed for the various ufuncs (including scatter-gather operations for strided arrays) and the variety of platforms NumPy supports.

**Comment by sturlamolden** [comment]
To use vector math libraries efficiently with strided or badly aligned data one would have to copy the data into a contiguous and aligned work array. For efficient use of cache it needs to be rather small.

**Comment by r-devulap** [comment]
@mattip I recommended making some changes which should improve performance further but looks good otherwise.

**Comment by mattip** [comment]
Thanks @xiegengxin

**Pull request merged**